### ISSUE TITLE
Query sharding should support vector selector

### ISSUE THREAD

**Issue by yeya24** [thanos-io/thanos#5605]
### Is your proposal related to a problem?

Currently query sharding is supported by the query frontend to shard both instant and range queries. However, only some special queries are shardable.

From my test, vector selector expressions are not shardable right now. For example, `up{job="prometheus"}` cannot be sharded because there is no sharding label. So this type of query cannot be benefited by query sharding.

### Describe the solution you'd like

Make this kind of expression shardable and shard series by hashmod (same as other type of expressions).

**Comment by GiedriusS** [comment]
How would we know according to which labels we should shard at the query-frontend level? :thinking:

**Comment by yeya24** [comment]
Maybe I missed something. Can we simply calculate hashmod based on the whole series labels? For example, query `http_requests_total{}` could match series
```
http_requests_total{code="200", handler="test"} hash mod % shards = 0
http_requests_total{code="400", handler="foo"} hash mod % shards = 1
http_requests_total{code="500", handler="bar"} hash mod % shards = 2
```

This is something in my mind and I think it is pretty straightforward. Is it on the right direction?

**Comment by fpetkovski** [comment]
That would work, but we should benchmark individual queries to see if performance will be better or worse. Since there is no  promql to execute, queriers will mostly be used as passthrough proxies. In any case, hard to say what the effect would be so it's worth trying it out.

**Comment by yeya24** [comment]
I agree @fpetkovski. We have to check the performance first.
But this is also important in the future once we support series sharding from the compactor side because at that time we can shard any metrics.

**Comment by alanprot** [comment]
I think any `over_timer` function can also be shardable even without `by` clause?

For example, for queries like `avg_over_time(kube_pod_labels[1h])` we can "shard by all labels"?

Note that `max(avg_over_time(kube_pod_labels[5m])) by (pod)` is shardable, so the above example should have the same performance benefits of this one?

**Comment by fpetkovski** [comment]
I see, that is indeed an interesting point. I would say it's worth giving it a shot 👍

**Comment by alanprot** [comment]
Yeah.. for sure need to run some tests.. but I think mostly of the benefit comes with the fact all this data transfer / process (specially btw query and storage nodes) will be done in parallel and distributed. Also, the memory footprint  on each of those nodes should be smaller.
